[docs] Add Recipes section

cherniavskii opened the pull request:
Adding the first recipe to kick off the discussion around this topic.
It was extracted from issues https://github.com/mui/mui-x/issues/2851 and https://github.com/mui/mui-x/issues/5007

Preview: https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/recipes-editing/

TODO:

- [x] Link from feature pages to appropriate Recipes sections
- [x] Move Editing "Advanced use-cases" to Recipes

Other candidates for recipes:
- [x] https://github.com/mui/mui-x/issues/5838
- [ ] https://github.com/mui/mui-x/issues/4248#issuecomment-1074928601
- [ ] https://github.com/mui/mui-x/issues/4127
- [ ] https://github.com/mui/mui-x/issues/5381
- [ ] https://github.com/mui/mui-x/issues/3656
- [ ] https://github.com/mui/mui-x/issues/2186#issuecomment-1189696795
- [ ] https://github.com/mui/mui-x/issues/5506#issuecomment-1189729426

cherniavskii commented on docs/data/pages.ts:
```diff
@@ -93,6 +93,9 @@ const pages: MuiPage[] = [
         ],
       },
+      {
+        pathname: '/x/react-data-grid/recipes',
```
We can do multiple pages as well (similar to `Advanced` category in the nav bar), but:
- those pages would be rather short
- it's easier to add new demo to existing page than create a new one

m4theushw replied:
Did you find any other example not related to editing? My concern is if we won't have a page that most recipes are related to editing. We already have https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/editing/#advanced-use-cases which are recipes. Maybe the sidebar should be:

Editing
- Fundamentals (goes to https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/editing/)
- Recipes (https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/editing/#advanced-use-cases + https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/recipes/)

Or we move everything from https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/editing/#advanced-use-cases to https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/recipes/

cherniavskii replied:
> Did you find any other example not related to editing?

A lot of them are indeed related to editing - like https://github.com/mui/mui-x/issues/5381
But we will probably have some not related to editing as well - like https://github.com/mui/mui-x/issues/4127 or https://github.com/mui/mui-x/issues/3656

+1 for moving "Advanced use cases" examples from "Editing" to "Recipes".

cherniavskii commented on docs/data/data-grid/recipes/recipes.md:
```diff
@@ -0,0 +1,15 @@
+
+### Multiline editing
+
+Creating custom component for inline editing of multiline text.
```
Not sure about the form - should it be more like a tutorial with step-by-step explanation or just a working demo that users can explore on their own?

m4theushw replied:
It depends. This specific demo uses only the principles already explained in https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/editing/#create-your-own-edit-component. Its "advanced" part is more of an advanced usage of MUI components.

alexfauquette replied:
For this specific demo I do not see how to split the explanation. It could contain one or two sentences about the big lines. Especially on the useof `'cellKeyDown'` which is not intuitive..

The strategy is to use a custom editing component that renders a `Popper` over the cell. In this proper you have enough space to use a multiline input.
Since the component is in the Popper, keyboard event will not be propagated to the grid. Then you need to publish the 'cellKeyDown' such that keyboard interaction like Ctrl+Enter or Escape quite the edit mode and saves or reverts the editing.

m4theushw replied:
Prefer active voice.

```suggestion
You can have columns with multiline text and edit them by creating a custom edit component.
```

m4theushw reviewed (commented):
We can also add https://github.com/mui/mui-x/issues/2186#issuecomment-1189696795 and https://github.com/mui/mui-x/issues/5506#issuecomment-1189729426

cherniavskii commented on docs/data/data-grid/recipes/recipes.md:
```diff
@@ -0,0 +1,15 @@
+
+### Multiline editing
+
+Creating custom component for inline editing of multiline text.
```
Yes, that was intentional to handle newline with "Enter" and "Enter + Shift".

cherniavskii commented on docs/data/data-grid/recipes/MultilineEditing.tsx:
```diff
@@ -0,0 +1,145 @@
+
+  const handleKeyDown = React.useCallback(
+    (event) => {
+      if (
+        event.key === 'Escape' ||
+        (event.key === 'Enter' &&
+          !event.shiftKey &&
+          (event.ctrlKey || event.metaKey))
+      ) {
+        console.log('publish events');
+        const params = apiRef.current.getCellParams(id, field);
+        apiRef.current.publishEvent('cellKeyDown', params, event);
+      }
```
I assume that `apiRef.current.publishEvent('cellKeyDown', params, event);` would handle Escape properly.
My goal was to avoid replicating this logic and delegate it to `cellKeyDown` handler in the editing API (this demo uses new editing API)

cherniavskii commented:
I've moved examples from Editing "Advanced use cases" section to Recipes page.
I've also added one more recipe for Row grouping: https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/recipes/#toggling-groups-on-row-click

m4theushw commented on docs/data/data-grid/editing/editing.md:
```diff
@@ -456,109 +456,7 @@ Instead, use the buttons available in each row or in the toolbar.
-You can change date format by importing different locale (`en-US` locale is used in the example above).
-See [Localization](/x/react-date-pickers/localization/) for more information.
-:::
+See [Recipes](/x/react-data-grid/recipes/#editing) for more advanced use cases.
```
Could you also create an advanced use cases in https://deploy-preview-5744--material-ui-x.netlify.app/x/react-data-grid/row-grouping/ and link to the recipes pages?

m4theushw reviewed (commented):
https://github.com/mui/mui-x/issues/5506#issuecomment-1189729426 can also be added to this page later. It's about creating an edit component with an Autocomplete.

flaviendelangle commented:
Sorry if the question has already been asked
But wouldn't it be better to already split the page in two (one for editing and one for row grouping) ?
To avoid a routing breaking change in a few months when this page will be too long

cherniavskii commented:
> Sorry if the question has already been asked But wouldn't it be better to already split the page in two (one for editing and one for row grouping) ? To avoid a routing breaking change in a few months when this page will be too long

I was wondering about this in https://github.com/mui/mui-x/pull/5744#discussion_r941655603
I agree that avoiding breaking URLs is a solid argument to split right now 👍

DanailH reviewed (approved):
It's a great first step. We can continuously add more recipes in the future.

The only feedback I have is that the code previews are not showing - it's probably an issue with the renaming of the file. Regenerating them should do the trick of fixing it.

oliviertassinari commented on docs/data/data-grid/recipes-editing/recipes-editing.md:
```diff
@@ -0,0 +1,118 @@
+title: Data Grid - Editing recipes
+---
+
+# Data grid - Editing recipes
```
It's great to have more content!

Now, are we sure these "features" are best documented in a misc section, called "Recipes"?

Looking at each demo, as a developer, I think that I would expect each of these features to be covered in side the "Editing" part of the docs. For example, AG Grid has many pages about editing: https://www.ag-grid.com/javascript-data-grid/undo-redo-edits/. It seems that we draw a line between, the feature is built-in vs. the feature should be composed but isn't a feature that needs to be composed as relevant as one that is built-in?

cherniavskii replied:
> Looking at each demo, as a developer, I think that I would expect each of these "features" to be covered in side the "Editing" part of the docs.

Adding more content to the page creates a different problem - pages are too long and we need to split them at some point. Offloading custom examples to Recipes reduces this problem.

> It seems that we draw a line between the feature is built-in vs. the feature is composed

For me it's more like "Essential" vs "Advanced".

> either the demo is in the docs, maintained like any other features, or it's not, it's something we quickly put together on a codesandbox with pinned dependencies

I think quick demo is still an option, but for popular customizations use cases it's better to have it maintained in the docs.
I feel like Recipes in this case fill the gap between codesandbox demos and docs - the demo is not important enough to increase the scope of the feature page, but is important enough to have it maintained.

oliviertassinari replied:
> Offloading custom examples to Recipes reduces this problem.

@cherniavskii I agree with this.
My question is more on: where should it be located (navigation tree + URL)?
I took the editing feature as an example. https://mui.com/x/react-data-grid/editing/ is one big page. If we use https://www.notion.so/mui-org/Technical-writing-7e55b517ac2e489a9ddb6d0f6dd765de#075e05fa0a934b59a5b0bded25dea935, then point 1. and 2. are asking us to break down the page. At this point, we might be able to move the content closer to the documentation of the feature. We can move https://mui.com/x/react-data-grid/recipes-editing/ to a recept editing page under the main "Editing" navigation 😁

> For me it's more like "Essential" vs "Advanced".

Fair enough.

> fill the gap between

Fair enough.

oliviertassinari replied:
So anyone interested/resonates with the idea of splitting https://mui.com/x/react-data-grid/editing/ into smaller pages 😁? Like @flaviendelangle did for #5195

oliviertassinari replied:
> As for pages that are not split yet, I assume we would just add Recipes section and if page grows too much then split?

@cherniavskii It sounds great.

> would it make sense to split and have another level of nesting

I don't know but it's not unheard of. AG Grid goes to 4 levels of depth with the docs of the data grid https://ag-grid.com/javascript-data-grid/grouping-single-group-column/. We go up to 3 levels. However in our case, the first level is the data grid component, in their case, they don't even have it. So it's like we were going 5 level deep, while we go down up to 3 right now. Personally, I think that AG Grid should flatten more

cherniavskii commented:
@DanailH 
> The only feedback I have is that the code previews are not showing - it's probably an issue with the renaming of the file. Regenerating them should do the trick of fixing it.

I turned previews off intentionally, because in case of Recipes preview is too short to be useful.
In the preview you'll only see the return data grid component and some props, while usually the important part is somewhere else - so you'll have to see the whole code anyway.

Qadir134 commented:
Has anyone solve the issue no: #5381 ?
